### PR TITLE
AMSgrad optimizer

### PR DESCRIPTION
**Pull request by kashif**

### CODE REVIEW

**Comment by ozabluda** [comment]
How difficult would it be to unify it with Adam? Maybe even `vhat*` can be calculated unconditionally (if performance difference is negligible) and only `p_t` calculation would differ between the two.

**Comment by kashif** [comment]
@ozabluda yes I was also thinking about that since we are sharing so much code... let me experiment more and see what others suggest...

**Comment by fchollet** [comment]
Maybe this could become an option in `Adam`?

**Comment by kashif** [comment]
sure @fchollet I could add it to all the appropriate methods RMSPROP, ADAM, ADADELTA,
NADAM, etc as the paper suggest...

What should we call the option and should it be on by default?

**Comment by fchollet** [comment]
> What should we call the option

Please make a few name suggestions!

> should it be on by default?

No, since this would modify the current behavior. We need to always be backwards compatible.

**Comment by kashif** [comment]
@fchollet ok ready for a review

File: keras/optimizers.py [hunk 1]
```diff
@@ -394,13 +394,15 @@ class Adam(Optimizer):
         beta_2: float, 0 < beta < 1. Generally close to 1.
         epsilon: float >= 0. Fuzz factor.
         decay: float >= 0. Learning rate decay over each update.
+        amsgrad: boolean. Weather to apply AMSGrad variant of this algorithm.
```
Comment (fchollet) [inline keras/optimizers.py]: Please mention the paper (ref below) in the arg description

File: keras/optimizers.py [hunk 2]
```diff
@@ -427,12 +430,18 @@ def get_updates(self, loss, params):
 
         ms = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
         vs = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
-        self.weights = [self.iterations] + ms + vs
+        vhats = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
```
Comment (fchollet) [inline keras/optimizers.py]: Only create these if `amsgrad=True` (they could be None if False)

**Comment by kashif** [comment]
@fchollet ok ready for another review. Should I add versions for the other methods or leave that for another PR?

**Comment by ozabluda** [comment]
FWIW, setting `amsgrad=True` in cifar10_resnet.py example, made test error to be 0.9208 for ResNet20v1, which is statistically the same as the documented 0.9216 in the example, which is approximately what I was getting earlier as well.

**Comment by kashif** [comment]
@ozabluda can you also kindly try on some RNN tasks?

**Comment by ozabluda** [comment]
FWIW, on a VGG-like CNN of mine, setting `amsgrad=True` improved test accuracy on CIFAR-10 from 0.9308 to 0.9326, which is approximately 0.7 stdev better, but with `amsgrad=True` but it was still improving at epoch 300, which doesn't happen with `amsgrad=False`, so this is when I stop.

@kashif, do you have a particular RNN in mind. There are no RNNs that use Adam in examples/. There are several that use RMSProp though.

**Comment by ahundt** [comment]
@ozabluda you might be able to do it with [save_all_weights()](https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/utils/save_load_utils.py#L10) in keras-contrib, which saves both the model and the optimizer weights.

**Comment by ozabluda** [comment]
FWIW, on a VGG-like CNN of mine, setting `amsgrad=True` causes test accuracy to slowly improve (not statistically significant) even at epoch 526, which is not happening with `amsgrad=True`. But I can't really tell if it's sorta improving (a bit more likely) or simply suppressed earlier random spike(s) (a bit less likely).

**Pull request merged**